codec-http2/src/main/java/io/netty/handler/codec/http2/ClearTextHttp2ServerUpgradeHandler.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http2;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerAdapter;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.codec.http.HttpServerCodec;
+import io.netty.handler.codec.http.HttpServerUpgradeHandler;
+import io.netty.handler.codec.http.HttpServerUpgradeHandler.UpgradeCodecFactory;
+
+import java.util.List;
+
+import static io.netty.buffer.Unpooled.unreleasableBuffer;
+import static io.netty.handler.codec.http2.Http2CodecUtil.connectionPrefaceBuf;
+
+/**
+ * Providing how starting http2 protocol, by h2c HTTP upgrade, or by h2 with Prior Knowledge.
+ * This handler will set up {@link PriorKnowlegeHandler}, {@link HttpServerCodec},
+ * and {@link UpgradeCodecFactory} into the pipeline when handler added.
+ * And will set up pipeline correctly once it detect the protocol is h2c or h2.
```

Comment:
these docs aren't that clear right now.  I think rather than explaining the implementation details, it's better to describe the outcome.  The part where we have the least details "will set up pipeline correctly" should be the focus of this javadoc.